### ISSUE TITLE
Do we need definitions for identifier and locator?

### ISSUE THREAD

**Issue by tcole3** [w3c/wpub#27]
Based on Monday's call and several issue threads (notably /wpub #11 and /publ-wg [21](https://github.com/w3c/publ-wg/issues/21)) we seem to have less than consensus about the difference between an **identifier** and a **locator**, and if there is a distinction, about whether the nature of this difference matters or is useful to highlight in the context of Web Publications (WP). Personally I think we need working definitions of identifier and of locator in WP terminology section. Below are my draft working definitions. But since it's unclear if the WG feels it necessary to distinguish between locator and identifier, before moving forward with a Pull Request, I would appreciate feedback about whether we need these definitions (as well as on the proposed wording). As a way to elaborate on the practical meaning of these definitions and to highlight the distinctions, I include comments below the proposed definitions. These comments are **not** intended for the terminology section of the spec (and might still need discussion), but may be helpful in drafting other sections.

**Identifier** 
An identifier **_persistently identifies_** a WP or WP resource. A WP or WP resource MAY be identified by more than one identifier, but an identifier MUST NOT identify more than one WP or WP resource. Different versions or editions of a WP { SHOULD? MAY? } have distinct identifiers. Some identifiers MAY also serve as locators. URIs [[rfc 3986]], IRIs[[rfc 3987]], URNs [[rfc 8141]], [DOIs](https://www.doi.org/), [ISBNs](https://www.isbn-international.org/), [PURLs ](https://archive.org/services/purl/) are all examples of identifiers frequently used in publishing.

**Locator**
A locator is a [URL ](https://www.w3.org/TR/html5/references.html#refsURL) that can be used to **_locate_** and **_retrieve_** a WP or WP resource, subject to authentication / authorization and similar access limits mediated by HTTP or other protocols. Locators MAY rely for retrieval functionality on HTTP redirection and/or intermediate pages, e.g., a "landing page." Unlike identifiers, the locators associated with a WP or WP resource MAY change.   Locators that persistently and unambiguously identify a single WP or WP resource are also identifiers. 

**_Comments about identifiers_**: Identifiers are **not required**, but generally { SHOULD? MAY} be assigned to a WP. An identifier once assigned may not be reused in the future to identify a different WP. A single identifier cannot identify both a part of a WP and the WP as a whole; it must identify only one or the other.  Some WPs and WP resources MAY be available in multiple representations (e.g., multiple serializations); an identifier identifies independent of representation. Some identifiers are also locators, but not all. Though URIs and URLs now share a common syntax, a distinction still made in this context (and in some other contexts) is that URIs identify while URLs locate. Other kinds of identifiers (e.g., DOIs) can be mapped (usually through services) to locators. Other kinds of identifiers (e.g., an ISBN) identify only, i.e., are not immediately useful for locating an online representation of a WP or WP component.  

**_Comments about locators_**: Locators are **required**. The serialization, formatting and even byte-level content of a representation retrieved when dereferencing a URL may be fixed or may depend on content negotiation, the client device used to resolve the URL, the client IP address and/or other HTTP headers accompanying the Request, e.g., an editor using a browser or other tool that provides a HTTP Authorization header MAY see a different view than an anonymous reader. In some serializations, the resource (e.g., manifest) requested using a locator MAY be embedded within and conflated with another resource, e.g., a manifest conflated within the HTML representation of a WP's primary resource.

**Comment by iherman** [comment]
I like this... thanks @tcole3!

We could (and we have the tendency) to go into a very detailed discussion trying to cross all the t-s, but I would propose we should not do that. In my view it is much more important to have an overall image for WP-s by having a FPWD soon (remember that we are supposed to publish a FPWD by the end of this year!). Trying to cover a large palette of issue. We will have another year at least to get everything crystal clear and perfect.

Bottom line: in my view, this text is great and we should have it incorporated into the editors' draft asap. Some fine details and points can be settled later.

**Comment by rdeltour** [comment]
I generally agree with @iherman that we shouldn't try to cross all the t-s; that said, a couple comments (with the intent to make it even simpler for the FPWD):

 * I don't even think we need a definition for **locator**. Why not just use **URL** throughout the document and be done with it? Maybe I'm missing a point…
 * For identifiers, at least for the FPWD, I would avoid making RFC2119 statements or use a strong definition that can't be enforced or verified:

> an identifier _persistently identifies_ a WP

→ this is a strong statement, which is a wishful declaration of intent rather than something our spec can enforce.

> an identifier MUST NOT identify more than one WP

→ again, the spec can't enforce this.

In fact, I'm not even sure that we need to define the term _identifier_ either.

If we really need a definition, I'm proposing instead:

**Identifier**
An _identifier_ is a piece of metadata that is intended to be used to refer to a WP [or WP resource??] in a persistent and unambiguous manner. DOIs, ISBNs, PURLs are all examples of identifiers frequently used in publishing.

**Comment by mattgarrish** [comment]
I'm somewhat concerned about adding RFC requirements to terminology. Typically requirements go in the body where the concepts apply, as readers aren't guaranteed to read over terminology.

**Comment by iherman** [comment]
If we indeed do some changes...

1. Reflecting on one note of @rdeltour: we agreed, on our last call, that we would rely on the terminology in the ["Publishing and Linking on the Web"](https://www.w3.org/TR/publishing-linking/#terminology) as much as we can. Maybe we should _not_ use the term "locator" in the first place, and use the term ["address"](https://www.w3.org/TR/publishing-linking/#dfn-address) which is defined in that note.

2. For the identifier: we already have a definition for a [WP Canonical Identifer](https://w3c.github.io/wpub/#dfn-web-publication-canonical-identifier). I wonder whether that text is not more the definition of an "identifier", and the canonical simply means "it is the preferred identifier". Again, trying to reduce the complexity...

**Comment by tcole3** [comment]
1. I am fine with shorter definitions and in particular with @rdeltour suggestion :

**Identifier**
An identifier is a piece of metadata that is intended to be used to refer to a WP [or WP resource??] in a persistent and unambiguous manner. DOIs, ISBNs, PURLs are all examples of identifiers frequently used in publishing.

2. I think we will use the term identifier in the WP spec (if not definition can be delete later). Identifiers are useful in common scenarios, e.g., see discussion in [section 3 of the ietf identifier link relation draft](https://tools.ietf.org/id/draft-vandesompel-identifier-00.html#rfc.section.3) (previously shared by @HadrienGardeur via email). For the many publishers who do assign identifiers, find them useful and do differentiate between identifier and url / locator / address, talking about identifiers in the WP spec will be a signal that the spec is relevant to them. Publishers that don't use identifiers can ignore. 

3. The phrase 'or WP resource' was an attempt to acknowledge that Primary and Secondary Resources and potentially even segments of these resources can be assigned persistent identity. Could leave out if judged unimportant, or could use the phrase 'Web Content' from ["Publishing and Linking on the Web"](https://www.w3.org/TR/publishing-linking/#terminology) to stay consistent with that document (see below), or I am open to something else.

4. Since use of identifiers will be a SHOULD or MAY, enforcement is not an issue, especially given the proposed rewording ('is intended to'). 

5. With regard to canonical identifier, I would go the other way. If we define identifier, I am not sure we need a definition of canonical identifier in our terminology section. If we have need in the spec to address the use of an identifier in conjunction with the canonical link relationship, the concept will be self-evident, and I'm not sure canonical will come up in the spec in any other context. 

6. A possible alternative definition to replace proposed locator definition, one that focuses on the unique concept of a WP locator / address (as distinct from a generic URL) and stays closer to the Terminology section of the Publishing and Linking Document which did define address and access:

**WP {address | locator}**
A URL with which a user agent can access a WP. The WP {address | locator} is not guaranteed persistent (though it MAY be). Using a {address | locator} to access a WP for display and/or other purposes may involve also retrieving (possibly recursively) additional web content which is embedded or included by reference.

**Comment by iherman** [comment]
>Identifier
>An identifier is metadata that is intended to be used to refer to a WP [or WP resource??] in a persistent and unambiguous manner. DOIs, ISBNs, PURLs are all examples of identifiers frequently used in publishing.
>

I think that, for our case (also in term of your remark for canonical identifier below) we should add the text that is in the current definition of a canonical identifier:

> It SHOULD be an IRI (and therefore also an address). If not an IRI, it MUST be possible to make a 1-to-1 mapping of the canonical identifier to a address.


>* I think we will use the term identifier in the WP spec (if not definition can be delete later). Identifiers are useful in common scenarios, e.g., see discussion in section 3 of the ietf identifier link relation draft (previously shared by @HadrienGardeur via email). For the many publishers who do assign identifiers, find them useful and do differentiate between identifier and url / locator / address, talking about identifiers in the WP spec will be a signal that the spec is relevant to them. Publishers that don't use identifiers can ignore.
>

+1

>* The phrase 'or WP resource' was an attempt to acknowledge that Primary and Secondary Resources and potentially even segments of these resources can be assigned persistent identity. Could leave out if judged unimportant, or could use the phrase 'Web Content' from "Publishing and Linking on the Web" to stay consistent with that document (see below), or I am open to something else.
>

I am actually not sure that we should go into this. An identifier on the WP level is important. Identifying a fragment or a resource can then be done via the identifier being an address. If it is not then it becomes messy anyway (would you assign a DOI to a paragraph?) and we do not really need it.

>* Since use of identifiers will be a SHOULD or MAY, enforcement is not an issue, especially given the proposed rewording ('is intended to').
>

+1

>* With regard to canonical identifier, I would go the other way. If we define identifier, I am not sure we need a definition of canonical identifier in our terminology section. If we have need in the spec to address the use of an identifier in conjunction with the canonical link relationship, the concept will be self-evident, and I'm not sure canonical will come up in the spec in any other context.

Agree with, additionally, what I noted above

>
>* A possible alternative definition to replace proposed locator definition, one that focuses on the unique concept of a WP locator / address (as distinct from a generic URL) and stays closer to the Terminology section of the Publishing and Linking Document which did define address and access:

>**WP {address | locator}**
>A URL with which a user agent can access a WP. The WP {address | locator} is not guaranteed persistent (though it MAY be). Using a {address | locator} to access a WP for display and/or other purposes may involve also retrieving (possibly recursively) additional web content which is embedded or included by reference.

I would do one step further. If we take the term **address** as a given by referring to the Publishing and Linking Together, the the definition of a WP address becomes simply:

A **WP address** is an **address** that refers to the WP as a whole, and it MUST enable the retrieval of a representation of the manifest of the Web Publication.

(Again taking over from the current definition.)

I would think you should make a PR with these, and we can declare it as GoodEnough for now:-)

**Comment by tcole3** [comment]
1. Consensus - Yes, working definition of WP identifier as well as of WP address (rather than locator) is a good idea for now, pending confirming this need based on references to WP identifier that come up elsewhere in the FPWD. 
2. Unclear is whether we need to define a Canonical WP identifier and how to define the term if definition needed, but that would be a separate discussion.
3. Given 1, we can close this issue.
